src/passes/TypeSSA.cpp

```diff
@@ -54,93 +54,37 @@
+// types.
+std::vector<HeapType> ensureTypesAreInNewRecGroup(std::vector<HeapType>&& types,
                                                   Module& wasm) {
-  auto num = recGroup.size();
-
-  std::vector<HeapType> types;
-  types.reserve(num);
-  for (auto type : recGroup) {
-    types.push_back(type);
+  std::unordered_set<RecGroup> existing;
+  for (auto type : ModuleUtils::collectHeapTypes(wasm)) {
+    existing.insert(type.getRecGroup());
   }
 
-  // Find all the heap types present before we create the new ones. The new
-  // types must not appear in |existingSet|.
-  std::vector<HeapType> existing = ModuleUtils::collectHeapTypes(wasm);
-  std::unordered_set<HeapType> existingSet(existing.begin(), existing.end());
-
-  // Check for a collision with an existing rec group. Note that it is enough to
-  // check one of the types: either the entire rec group gets merged, so they
-  // are all merged, or not.
-  if (existingSet.count(types[0])) {
-    // Unfortunately there is a conflict. Handle it by adding a "hash" - a
-    // "random" extra item in the rec group that is so outlandish it will
-    // surely (?) never collide with anything. We must loop while doing so,
-    // until we find a hash that does not collide.
-    //
-    // Note that we use uint64_t here, and deterministic_hash_combine below, to
-    // ensure our output is fully deterministic - the types we add here are
-    // observable in the output.
-    uint64_t hashSize = num + 10;
-    uint64_t random = num;
-    while (1) {
-      // Make a builder and add a slot for the hash.
-      TypeBuilder builder(num + 1);
-      for (Index i = 0; i < num; i++) {
-        builder[i].copy(types[i]);
-      }
-
-      // Implement the hash as a struct with "random" fields, and add it.
-      Struct hashStruct;
-      for (Index i = 0; i < hashSize; i++) {
-        // TODO: a denser encoding?
-        auto type = (random & 1) ? Type::i32 : Type::f64;
-        deterministic_hash_combine(random, hashSize + i);
-        hashStruct.fields.push_back(Field(type, Mutable));
-      }
-      builder[num] = hashStruct;
-
-      // Build and hope for the best.
-      builder.createRecGroup(0, num + 1);
-      auto result = builder.build();
-      assert(!result.getError());
-      types = *result;
-      assert(types.size() == num + 1);
-
-      if (existingSet.count(types[0])) {
-        // There is still a collision. Exponentially use larger hashes to
-        // quickly find one that works. Note that we also use different
-        // pseudorandom values while doing so in the for-loop above.
-        hashSize *= 2;
-      } else {
-        // Success! Leave the loop.
-        break;
-      }
-    }
+  UniqueRecGroups unique(wasm.features);
+  for (auto group : existing) {
+    std::vector<HeapType> types(group.begin(), group.end());
+    [[maybe_unused]] auto uniqueTypes = unique.get(std::move(types));
+    assert(uniqueTypes.size() == group.size() && "unexpected collision");
```

Comment:
The fuzzer found a way to make this fail after about 80k iterations, so I'll investigate and fix that as well.

Reply:
This can still happen when custom descriptors are disabled and the _input_ already includes rec groups that differ only in exactness. This is nonsensical, but our validator doesn't reject it, so the fuzzer can still generate it. I'll investigate rejecting these cases in the parsers.